### PR TITLE
[FLINK-1952] [jobmanager] Rework and fix slot sharing scheduler

### PR DESCRIPTION
**Pull request by StephanEwen**

### CODE REVIEW

**Comment by StephanEwen** [comment]
The latest commit "Add big not so mini cluster test for CC to provoke scheduler problem" is not going to be merged. It is solely for verifying that the scheduler now correctly handles jobs with iterations and higher parallelism and many TaskManagers.

You can start a 100 TaskManager cluster with that test setup in the commit and run connected components. (Give the VM 5GB heap space, then it works smoothly).

**Comment by StephanEwen** [comment]
Sorry for the whitespace re-formatting. My IntelliJ settings got reverted somehow, so it started auto-reformatting code :-/

File: flink-runtime/src/main/java/org/apache/flink/runtime/instance/Slot.java [hunk 1]
```diff
@@ -24,27 +24,40 @@
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 
 /**
- * Base class for task slots. TaskManagers offer one or more task slots, they define how many
- * parallel tasks or task groups a TaskManager executes.
+ * Base class for task slots. TaskManagers offer one or more task slots, which define a slice of 
+ * their resources.
+ *
+ * <p>In the simplest case, a slot holds a single task ({@link SimpleSlot}). In the more complex
+ * case, a slot is shared ({@link SharedSlot}) abd contains a set of tasks. Shared slots may contain
```
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/instance/Slot.java]: typo: abd -> and

File: flink-runtime/src/test/java/org/apache/flink/runtime/instance/SharedSlotsTest.java [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+
+			assertEquals(0, assignment.getNumberOfSlots());
+			assertEquals(0, assignment.getNumberOfAvailableSlotsForGroup(vertexId));
+			
+			// we should not be able to allocate any children from this released slot
+			assertNull(slot.allocateSharedSlot(new AbstractID()));
+			assertNull(slot.allocateSubSlot(new AbstractID()));
+			
+			// we cannot add this slot to the assignment group
+			assignment.addSharedSlotAndAllocateSubSlot(slot, Locality.NON_LOCAL, vertexId);
```
Comment (tillrohrmann) [inline flink-runtime/src/test/java/org/apache/flink/runtime/instance/SharedSlotsTest.java]: Is an assertion which checks that the slot was not added missing here?

File: flink-tests/src/test/resources/log4j.properties [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# A1 is set to be a ConsoleAppender.
+log4j.appender.testlogger=org.apache.log4j.ConsoleAppender
+log4j.appender.testlogger.target = System.err
+log4j.appender.testlogger.layout=org.apache.log4j.PatternLayout
+log4j.appender.testlogger.layout.ConversionPattern=%-4r [%t] %-5p %c %x - %m%n
+
+log4j.logger.org.apache.flink.runtime.client.JobClient=OFF
+
+# suppress the irrelevant (wrong) warnings from the netty channel handler
+log4j.logger.org.jboss.netty.channel.DefaultChannelPipeline=ERROR, testlogger
```
Comment (tillrohrmann) [inline flink-tests/src/test/resources/log4j.properties]: Was this file added for debugging purposes?

**Comment by tillrohrmann** [comment]
LGTM.

**Comment by StephanEwen** [comment]
Thanks!

I'll address the comments, remove the accidentally added `log4j.properties` file and merge this.

**Pull request closed without merging**